torch/_functorch/_aot_autograd/autograd_cache.py

```diff
@@ -164,8 +222,297 @@ def autograd_cache_hash(
+        return result
+
+
+@dataclass
+class CompiledForward(FXGraphCacheLoadable):
+    """
+    Cacheable entry for a forward function
+    """
+
+    pass
+
+
+@dataclass
+class CompiledBackward(FXGraphCacheLoadable):
+    """
+    Cacheable entry for a forward function
+    """
+
+    # Used by AOTDispatchAutograd.post_compile
+    backward_state_indices: List[int]
+    num_symints_saved_for_bw_: int
+
+
+@dataclass
+class AOTAutogradCacheEntry:
+    """A single entry into the cache."""
+
+    # Forward and Backward info
+    compiled_fw: CompiledForward
+    compiled_bw: Optional[CompiledBackward]
+
+    # Runtime_metadata saved right before compilation
+    runtime_metadata: ViewAndMutationMeta
+
+    # Wrappers that run after each aot_dispatch_* function
+    dispatch_wrappers: List[CompilerWrapper]
+
+    # Used by AOTSubclassWrapper
+    maybe_subclass_meta: Optional[SubclassMeta]
+    num_fw_outs_saved_for_bw: Optional[int]
+
+    # Used by RuntimeWrapepr
+    indices_of_inps_to_detach: List[int]
+
+    # Turn cache entry into the original callable
+    def wrap_post_compile(
+        self, args: List[torch.Tensor], aot_config: AOTConfig
+    ) -> Callable:
+        compiled_fw_func = self.compiled_fw.load(args)
+        compiled_bw_func = None
+        if self.compiled_bw is not None:
+            compiled_bw_func = self.compiled_bw.load(args)
+            needs_autograd = True
+        else:
+            needs_autograd = False
+
+        # Wrap the forward function in post compile wrappers
+        compiled_fw_func = AOTDispatchSubclassWrapper(
+            trace_joint=needs_autograd,
+            fw_only=None,
+            maybe_subclass_meta=self.maybe_subclass_meta,
+            num_fw_outs_saved_for_bw=self.num_fw_outs_saved_for_bw,
+        ).post_compile(
+            compiled_fw_func, aot_config, runtime_metadata=self.runtime_metadata
+        )
+
+        # In autograd case, functionalizedRngWrapper should not modify outs
+        return_new_outs = not needs_autograd
+        compiled_fw_func = FunctionalizedRngRuntimeWrapper(
+            return_new_outs=return_new_outs
+        ).post_compile(
+            compiled_fw_func, aot_config, runtime_metadata=self.runtime_metadata
+        )
+        disable_amp = torch._C._is_any_autocast_enabled()
+
+        if needs_autograd:
+            assert self.compiled_bw is not None
+            compiled_function = AOTDispatchAutograd.post_compile(
+                compiled_fw_func,
+                compiled_bw_func,
+                self.maybe_subclass_meta,
+                self.compiled_bw.num_symints_saved_for_bw_,
+                self.compiled_bw.backward_state_indices,
+                disable_amp,
+                self.indices_of_inps_to_detach,
+                None,  # lazy_backward_info
+                aot_config,
+                fw_metadata=self.runtime_metadata,
+            )
+        else:
+            compiled_function = RuntimeWrapper(
```

Comment:
I needed to remind myself why `RuntimeWrapper.post_compile` is only getting called in one of these two branches - it's because:

- `RuntimeWrapper.post_compile` handles the input mutations for both inference and training (although it does some training-specific things too, like detaching inputs, setting grad state, replaying views)

- `AOTDispatchAutograd.post_compile` implicitly instantiates `RuntimeWrapper.post_compile` (so we still get it in the training path)

The runtime wrappers are just... pretty complicated (and the explicit types for each wrapper are a strict improvement). I'm just thinking about any ways to make this code more readable. Maybe it's worth adding a comment explaining which wrappers are called here?

Reply:
I think it might be worth refactoring aot_dispatch_base into a AOTDispatchBase.post_compile function too, so that I can just split between two big post compile calls. It also shares the code in runtime_wrappers so there's less deviation between the two.